### ISSUE TITLE
straight.el bootstrapping slows down emacs startup considerably

### ISSUE THREAD

**Issue by jkp** [raxod502/straight.el#373]
Hi

My startup was feeling sluggish so I used `esup` to profile my init file to try to work out what was slowing things down. I got the following surprising results:

    Total User Startup Time: 0.804sec     Total Number of GC Pauses: 1     Total GC Time: 0.020sec

    bootstrap.el:86  0.465sec   57%
    (straight-use-recipes '(melpa :type git :host github
    :repo "melpa/melpa"
    :no-build t))

    bootstrap.el:100  0.250sec   31%
    (straight-use-recipes '(emacsmirror :type git :host github
    :repo "emacsmirror/epkgs"
    :nonrecursive t
    :no-build t))
    
    init.el:153  0.020sec   2%
    (setq gc-cons-threshold (* 2 1000 1000))

I looked at `bootstrap.el` and found that there are no switches to control whether these functions are called or not: I'm interested if anyone else sees similar results / if there is something I'm doing wrong?  I'm using the standard bootstrapping snippet as described in the documentation.

Thanks

**Comment by jkp** [comment]
Digging through the open issues I found an [old issue](https://github.com/raxod502/straight.el/issues/41) that had some ideas of things to try.  I customised `traight-check-for-modifications ` as suggested [here](https://github.com/raxod502/straight.el/issues/41#issuecomment-453747530) and it brought the time right down.

I don't know your code well enough to understand the tradeoffs of this tweak: can you enlighten me?

Thanks for an awesome package manager!

**Comment by raxod502** [comment]
Aside from the information available in the user manual and documentation strings, here is a summary of the tradeoffs:

### `find-at-startup`

Save build timestamps and run `find(1)` at startup to detect changes

* Most reliable, never misses changes
* Requires `find(1)`
* Slows down startup

### `check-on-save`

Use `before-save-hook` to detect changes

* No external dependencies
* No startup delay
* No additional CPU or memory impact
* Doesn't catch changes made except via `save-file` inside Emacs

### `watch-files`

Run filesystem watcher to detect changes

* Requires Python 3 and Watchexec
* No startup delay
* Takes a few seconds to build virtualenv the first time
* Memory and CPU impact of running filesystem watcher
* Only misses changes if you make them after booting the system but before starting Emacs

---

Do you have any suggestions on how the documentation could be improved so you would have had an easier time finding this out?

**Comment by DoMiNeLa10** [comment]
I don't know whether this is the right place to mention this, as I'm not sure whether I'm doing something wrong or this is a genuine bug, but even when I set `straight-check-for-modifications` to `(check-on-save)`, `straight.el` still runs `find`. This slows down startup considerably, especially on machines with mechanical hard drives. Emacs startup time became excruciatingly slow for me after switching to this package manager, and it seems that no matter what I set as the value for that variable, `find` ends up running.

**Comment by raxod502** [comment]
@DoMiNeLa10 Have you made sure to set that variable _before_ bootstrapping `straight.el`?

**Comment by DoMiNeLa10** [comment]
I have, I realized that's the way I should've been doing it recently, and no matter what I set there, `find` still runs on startup. [Here's the file where I initialize straight.el](https://github.com/DoMiNeLa10/.emacs.d/blob/473256627af35de5e94b8da209afc7ddceb94b7e/config/my-straight-init.el). I've tried plenty of options including `nil`, which should disable checking completely.

**Comment by raxod502** [comment]
@DoMiNeLa10 I cloned your repository into my `~/.emacs.d` and cannot see any `find` command being run in `*straight-process*` apart from the version-checking one. (I suppose we could make that lazy, so that it's only run if we actually do need to run a `find` command, but that shouldn't be causing a problem.)

Feel free to open a new issue with more precise directions to reproduce, if you still experience this issue.

**Comment by priyadarshan** [comment]
> Do you have any suggestions on how the documentation could be improved so you would have had an easier time finding this out?

It would be nice to have that summary in the documentation, perhaps at the end of [the related section](https://github.com/raxod502/straight.el/blob/c42edf91249a3610a872402082e977c61cf65dc9/README.md#customizing-when-packages-are-built)?

It is all already there, but having it as list makes it quite clearer.

**Comment by raxod502** [comment]
As you wish :)

**Comment by priyadarshan** [comment]
Thank you, O straight.el's good genie :)

**Comment by raxod502** [comment]
This thread is being closed automatically by [Tidier](https://github.com/raxod502/tidier) because it is labeled with "waiting on response" and has not seen any activity for 90 days. But don't worry—if you have any information that might advance the discussion, leave a comment and the thread may be reopened :)
